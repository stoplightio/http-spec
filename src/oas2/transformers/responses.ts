import { isPlainObject } from '@stoplight/json';
import type { IHttpOperationResponse, Optional } from '@stoplight/types';
import { DeepPartial } from '@stoplight/types';
import { Operation } from 'swagger-schema-official';

import { isNonNullable } from '../../guards';
import { translateSchemaObject } from '../../oas/transformers/schema';
import { entries } from '../../utils';
import { getExamplesFromSchema, getProduces } from '../accessors';
import { isResponseObject } from '../guards';
import { Oas2TranslateFunction } from '../types';
import { translateToHeaderParams } from './params';

const translateToResponse: Oas2TranslateFunction<
  [produces: string[], statusCode: string, response: unknown],
  Optional<IHttpOperationResponse>
> = function (produces, statusCode, response) {
  const resolvedResponse = this.maybeResolveLocalRef(response);
  if (!isResponseObject(resolvedResponse)) return;

  this.state.enter('responses', statusCode);

  const headers = translateToHeaderParams.call(this, resolvedResponse.headers);
  const objectifiedExamples = entries(
    resolvedResponse.examples || (resolvedResponse.schema ? getExamplesFromSchema(resolvedResponse.schema) : void 0),
  ).map(([key, value]) => ({ id: this.generateId('example'), key, value }));

  const contents = produces
    .map(produceElement => ({
      id: this.generateId('security'),
      mediaType: produceElement,
      schema: isPlainObject(resolvedResponse.schema)
        ? translateSchemaObject.call(this, resolvedResponse.schema)
        : void 0,
      examples: objectifiedExamples.filter(example => example.key === produceElement),
    }))
    .filter(({ schema, examples }) => !!schema || examples.length > 0);

  const translatedResponses = {
    id: this.generateId('security'),
    code: statusCode,
    description: resolvedResponse.description,
    headers,
    contents,
  };

  const foreignExamples = objectifiedExamples.filter(example => !produces.includes(example.key));
  if (foreignExamples.length > 0) {
    if (translatedResponses.contents.length === 0)
      translatedResponses.contents[0] = {
        id: this.generateId('param'),
        mediaType: '',
        schema: {},
        examples: [],
      };

    translatedResponses.contents[0].examples!.push(...foreignExamples);
  }

  return translatedResponses;
};

export const translateToResponses: Oas2TranslateFunction<
  [operation: DeepPartial<Operation>],
  IHttpOperationResponse[]
> = function (operation) {
  const produces = getProduces(this.document, operation);
  return entries(operation.responses)
    .map(([statusCode, response]) => translateToResponse.call(this, produces, statusCode, response))
    .filter(isNonNullable);
};
