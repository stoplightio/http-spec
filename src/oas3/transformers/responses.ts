import { Dictionary, IHttpHeaderParam, IHttpOperationResponse, IMediaTypeContent, Optional } from '@stoplight/types';
import { compact, map } from 'lodash';
import { ResponseObject, ResponsesObject } from 'openapi3-ts';

import { translateHeaderObject, translateMediaTypeObject } from './content';

function translateToResponse(response: ResponseObject, statusCode: string): IHttpOperationResponse {
  return {
    code: statusCode,
    description: response.description,
    headers: compact<IHttpHeaderParam>(
      map<Dictionary<unknown> & unknown, Optional<IHttpHeaderParam>>(response.headers, translateHeaderObject),
    ),
    contents: compact<IMediaTypeContent>(
      map<Dictionary<unknown> & unknown, Optional<IMediaTypeContent>>(response.content, translateMediaTypeObject),
    ),
  };
}

export function translateToResponses(
  responses: ResponsesObject,
): IHttpOperationResponse[] & { 0: IHttpOperationResponse } {
  return map<ResponsesObject, IHttpOperationResponse>(responses, translateToResponse) as IHttpOperationResponse[] & {
    0: IHttpOperationResponse;
  };
}
