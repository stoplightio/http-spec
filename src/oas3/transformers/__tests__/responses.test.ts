import type { DeepPartial } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';

import { translateToResponses } from '../responses';

describe('translateToOas3Responses', () => {
  it('given empty dictionary should return empty array', () => {
    expect(translateToResponses({}, {})).toEqual([]);
  });

  it('given a response in dictionary should translate', () => {
    expect(
      translateToResponses(
        {},
        {
          default: {
            content: {
              'fake-content-type': {},
            },
            description: 'descr',
            headers: {
              'fake-header-name-1': {
                description: 'calls per hour allowed by the user',
                schema: {
                  type: 'integer',
                  format: 'int32',
                },
                example: 1000,
              },
              'fake-header-name-2': {
                description: 'calls per hour allowed by the user',
                schema: {
                  type: 'integer',
                  format: 'int32',
                },
                required: true,
                example: 1000,
              },
            },
          },
          200: {
            content: {
              'fake-content-type-200': {
                example: 'dumb',
              },
            },
            description: 'descr 200',
            headers: {
              'fake-header-name-200': {},
            },
          },
        },
      ),
    ).toMatchSnapshot();
  });

  it('given a response with nullish headers in dictionary should translate', () => {
    expect(
      translateToResponses(
        {},
        {
          200: {
            headers: {
              '0': null,
            },
          },
        },
      ),
    ).toStrictEqual([
      {
        code: '200',
        contents: [],
        description: void 0,
        headers: [],
      },
    ]);
  });

  it('should skip nullish responses', () => {
    expect(
      translateToResponses(
        {},
        {
          200: null,
          201: {
            description: 'description 201',
          },
        },
      ),
    ).toStrictEqual([
      {
        code: '201',
        contents: [],
        description: 'description 201',
        headers: [],
      },
    ]);
  });

  it('should resolve $refs nullish responses', () => {
    const document: DeepPartial<OpenAPIObject> = {
      openapi: '3.0.0',
      paths: {
        '/user': {
          get: {
            responses: {
              '200': {
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/User',
                    },
                    examples: {
                      'my-example': {
                        $ref: '#/components/examples/Joe',
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
      components: {
        schemas: {
          User: {
            title: 'User',
            type: 'object',
            properties: {
              id: {
                type: 'integer',
              },
            },
          },
        },
        examples: {
          Joe: {
            value: {
              id: 1,
            },
          },
        },
      },
    };

    expect(translateToResponses(document, document.paths!['/user'].get.responses)).toEqual([
      {
        code: '200',
        contents: [
          {
            encodings: [],
            examples: [
              {
                key: 'my-example',
                value: {
                  id: 1,
                },
              },
            ],
            mediaType: 'application/json',
            schema: {
              $schema: 'http://json-schema.org/draft-07/schema#',
              properties: {
                id: {
                  type: 'integer',
                },
              },
              title: 'User',
              type: 'object',
            },
          },
        ],
        headers: [],
      },
    ]);
  });
});
