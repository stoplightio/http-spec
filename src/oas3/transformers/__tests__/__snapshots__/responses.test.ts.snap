// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateToOas3Responses given a response in dictionary should translate 1`] = `
Array [
  Object {
    "code": "200",
    "contents": Array [
      Object {
        "encodings": Array [],
        "examples": Array [
          Object {
            "key": "default",
            "value": "dumb",
          },
        ],
        "mediaType": "fake-content-type-200",
        "schema": undefined,
      },
    ],
    "description": "descr 200",
    "headers": Array [
      Object {
        "encodings": Array [],
        "examples": Array [],
        "name": "fake-header-name-200",
        "style": "simple",
      },
    ],
  },
  Object {
    "code": "default",
    "contents": Array [],
    "description": "descr",
    "headers": Array [
      Object {
        "description": "calls per hour allowed by the user",
        "encodings": Array [],
        "examples": Array [
          Object {
            "key": "__default",
            "value": 1000,
          },
        ],
        "name": "fake-header-name-1",
        "schema": Object {
          "format": "int32",
          "type": "integer",
        },
        "style": "simple",
      },
      Object {
        "description": "calls per hour allowed by the user",
        "encodings": Array [],
        "examples": Array [
          Object {
            "key": "__default",
            "value": 1000,
          },
        ],
        "name": "fake-header-name-2",
        "required": true,
        "schema": Object {
          "format": "int32",
          "type": "integer",
        },
        "style": "simple",
      },
    ],
  },
]
`;
